grpc/src/main/java/io/stargate/grpc/service/interceptors/NewConnectionInterceptor.java

```diff
@@ -0,0 +1,128 @@
+package io.stargate.grpc.service.interceptors;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import io.grpc.Context;
+import io.grpc.Contexts;
+import io.grpc.Grpc;
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+import io.stargate.auth.AuthenticationService;
+import io.stargate.auth.AuthenticationSubject;
+import io.stargate.auth.UnauthorizedException;
+import io.stargate.db.AuthenticatedUser;
+import io.stargate.db.ClientInfo;
+import io.stargate.db.Persistence;
+import io.stargate.db.Persistence.Connection;
+import io.stargate.grpc.service.GrpcService;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.immutables.value.Value;
+
+public class NewConnectionInterceptor implements ServerInterceptor {
+
+  public static final Metadata.Key<String> TOKEN_KEY =
+      Metadata.Key.of("X-Cassandra-Token", Metadata.ASCII_STRING_MARSHALLER);
+
+  private static final InetSocketAddress DUMMY_ADDRESS = new InetSocketAddress(9042);
+
+  private final Persistence persistence;
+  private final AuthenticationService authenticationService;
+  private final LoadingCache<RequestInfo, Connection> connectionCache =
+      Caffeine.newBuilder().expireAfterWrite(Duration.ofMinutes(1)).build(this::newConnection);
+
+  @Value.Immutable
+  interface RequestInfo {
+
+    String token();
+
+    Map<String, String> headers();
+
+    @Nullable
+    SocketAddress remoteAddress();
+  }
+
+  public NewConnectionInterceptor(
+      Persistence persistence, AuthenticationService authenticationService) {
+    this.persistence = persistence;
+    this.authenticationService = authenticationService;
+  }
+
+  @Override
+  public <ReqT, RespT> Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> call, Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+    try {
+      String token = headers.get(TOKEN_KEY);
+      if (token == null) {
+        call.close(Status.UNAUTHENTICATED.withDescription("No token provided"), new Metadata());
+        return new NopListener<>();
+      }
+
+      RequestInfo info =
+          ImmutableRequestInfo.builder()
+              .token(token)
+              .headers(convertAndFilterHeaders(headers))
+              .remoteAddress(call.getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR))
+              .build();
+
+      Connection connection = connectionCache.get(info);
+
+      Context context = Context.current();
+      context = context.withValue(GrpcService.CONNECTION_KEY, connection);
+      return Contexts.interceptCall(context, call, headers, next);
+    } catch (Exception e) {
```

Comment:
can some other exception be thrown here, so we report falsely the `UNAUTHENTICATED`? Maybe we can split the catch into multiple ones with correct mapping?